FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/FhirToDataLakeOrchestratorJob.cs

```diff
@@ -382,83 +387,110 @@ private async Task<List<PatientWrapper>> GetToBeProcessedPatientsAsync(Cancellat
-                        _logger.LogInformation("Operation cancelled by customer.");
-                        throw new OperationCanceledException("Operation cancelled by customer.");
-                    }
-
-                    completedJobIds.Add(latestJobInfo.Id);
                 }
+                else if (firstRunningJobInfo.Status == JobStatus.Failed)
+                {
+                    _logger.LogInformation("The processing job is failed.");
+                    throw new RetriableJobException("The processing job is failed.");
+                }
+                else if (firstRunningJobInfo.Status == JobStatus.Cancelled)
+                {
+                    _logger.LogInformation("Operation cancelled by customer.");
+                    throw new OperationCanceledException("Operation cancelled by customer.");
+                }
+
+                UpdateRunningJobStatus(firstRunningJobId);
+                progress.Report(JsonConvert.SerializeObject(_result));
             }
 
-            if (completedJobIds.Count > 0)
+            _logger.LogInformation($"Orchestrator job {_jobInfo.Id} finished checking the first running job {firstRunningJobId} status, the job {firstRunningJobId} is {((JobStatus)firstRunningJobInfo.Status).ToString("D")}.");
+        }
+
+        private long GetFirstRunningJobId()
+        {
+            return _inputData.JobVersion switch
             {
-                _result.RunningJobIds.ExceptWith(completedJobIds);
-                progress.Report(JsonConvert.SerializeObject(_result));
+                SupportedJobVersion.V1 or SupportedJobVersion.V2 => _result.RunningJobIds.Min(),
+                SupportedJobVersion.V3 => _result.SequnceIdToJobIdMap[_result.CompletedJobCount],
+                _ => throw new SynapsePipelineInternalException($"The job version {_inputData.JobVersion} is unsupported."),
+            };
+        }
+
+        private void UpdateRunningJobStatus(long jobId)
+        {
+            switch (_inputData.JobVersion)
+            {
+                case SupportedJobVersion.V1:
+                case SupportedJobVersion.V2:
+                    _result.RunningJobIds.Remove(jobId);
+                    break;
+                case SupportedJobVersion.V3:
+                    _result.CompletedJobCount++;
+                    break;
+                default:
+                    throw new SynapsePipelineInternalException($"The job version {_inputData.JobVersion} is unsupported.");
             }
+        }
 
-            _logger.LogInformation($"Orchestrator job {_jobInfo.Id} finished checking running job status, there are {completedJobIds.Count} jobs completed.");
+        private int GetRunningJobCount()
+        {
+            return _inputData.JobVersion switch
+            {
+                SupportedJobVersion.V1 or SupportedJobVersion.V2 => _result.RunningJobIds.Count,
+                SupportedJobVersion.V3 => (int)(_result.CreatedJobCount - _result.CompletedJobCount),
```

Comment:
Seems this will count the failed retriable job, but those job are not actually been processing (waiting for processing job take them again).
Does **RunningJobIds** also count them? Do we expect this behavior?

Reply:
Yes, the `running jobs` here means the jobs that are `created while not completed`.